### ISSUE TITLE
Shebang first line should be excluded from maximum line length

### ISSUE THREAD

**Issue by robjens** [eslint/eslint#5633]
Having the following error message which seems to be correct, and expected, but arguing for a slight change of policy here:

`src/server/database.js|1 col 1 error| Line 1 exceeds the maximum line length of 100. (max-len) `

I don't think shebangs should count when checking maximum line length. The motivation for this is given the extreme difficulties (polyglot) multi-line shebangs bring to the table, add to that the common use practice of having shell executable node.js programs, one is inclined to think a configuration such as below, is a viable form of opening a JS file (this is what actually triggered the error).

```js
#!/usr/bin/env NODE_PATH='.' NODE_ENV='development' NODE_CONFIG_DIR='./config' /usr/local/bin/babel-node
```

The point regarding this is backed even further (I feel) by the fact that a lot of these so-called environment variables, usually found in either system-, or user-wide shell scope really don't apply so much to quite a few Node.js variables (or how they are used). E.g. I can have plenty of projects all with different env a few with development, others with production. Same goes for config.. (ok doesn't make for very good practice to keep changing around the config path, but you get my drift).

Now it is also considered bad practice to set environment variables from (JS) code, e.g. `process.env.NODE_PATH = '.'` (doesn't work actually, too late for that but the following does) or `process.env.NODE_ENV = 'development';`.

From this, I've come to conclude in a most of our use cases, the shebangs are really the most viable solution to cleanly and reliably set these conditions in quickly to run one-off scripts, for testing, debugging, development etc. (the production services all rely on systemd service environment  variables).

It goes without saying that I rather not set the whole eslint to allow for e.g. 120 columns or beyond.

Thoughts?

**Comment by robjens** [comment]
On second thought, one might also consider the alternative viewpoint that these are in their very nature, temporal and might fall under the same category as `no-console` and `spaced-comment` in all development scripts. That still leaves those that are actually intended to be used as `bin/program` shell script wrappers/starters although I guess the removal of extension does resolve (all) errors in the end =/

Anyone more capable than me to judge can close or chip in... thanks

**Comment by kaicataldo** [comment]
Hey there! This feels to me like a really niche use-case. max-len ignoring only shebang comments feels to me like unnecessary (and confusing) complexity - personally, as a user, that's not the behavior I'd expect.

Have you considered using the `comments` option in the [max-len docs](http://eslint.org/docs/rules/max-len)? ESLint also has the ability to ignore a line with `eslint-disable-` comments. An example of this use case would be:

```js
// eslint-disable-next-line max-len
#!/usr/bin/env NODE_PATH='.' NODE_ENV='development' NODE_CONFIG_DIR='./config' /usr/local/bin/babel-node
```

I'll let others chip in too, but your case seems like the very situation the `eslint-disable-` comments were created for - niche (but valid) edge cases!

**Comment by michaelficarra** [comment]
@kaicataldo The shebang has to be on the first line. I'm in favour of an option in `max-len` to ignore shebangs.

**Comment by rhalff** [comment]
I'm having a similar issue with the `spaced-comment` rule:

```javascript
$ cat shebang.js 
#!/usr/bin/env node
```

```javascript
$ eslint --no-eslintrc --rule '{"spaced-comment": 2}' shebang.js 

/home/rhalff/shebang.js
  1:1  error  Expected space or tab after '//' in comment  spaced-comment
```

Perhaps eslint itself could just ignore the shebang?

The line is not related to javascript and should therefor not be parsed by any rules.

Else if that's considered too expensive have the affected rules test for a shebang whenever an error occurs and let them ignore their own error.

**Comment by kaicataldo** [comment]
Closing this because we haven't moved forward with it in months.

**Comment by shreevatsa** [comment]
There is `ignore-pattern` which seems like it should work for this.

**Comment by robjens** [comment]
Ah... thanks for the tip shreevatsa.
